### ISSUE TITLE
QA

### ISSUE THREAD

**Issue by fliem** [BIDS-Apps/freesurfer#9]
Is there anything I can do to revive the QA branch?

**Comment by chrisgorgo** [comment]
It requires using a mini X server and I could not get it to work reliably under Docker and Singularity. It's an old and clunky solution - are there any other QA oriented FreeSurfer tools out there?

**Comment by fliem** [comment]
Mhh. I have a vague memory of @satra heading a past brainhack project about FS QA?

**Comment by satra** [comment]
@fliem - this depends on whether you are willing to create something from scratch or use some of the outputs in place.

for example, the brain is generally quite symmetric - so bad reconstructions often have `abs(1 - L_struct/R_struct) > some number` for each aparc/a2009 structure

but the things that need QA and are hard are:

- meninges part of cortical ribbon
- intensity bias affecting placement of surfaces in orbitofrontal cortex
- ripples caused by interaction between motion and fat shifts.
- outlines of subcortical boundaries (e.g., amygdala)

many of these should be detectable with enough training data.

**Comment by fliem** [comment]
Since surface plots now arrived in nilearn, and creating qc plots for freesurfer is such a drag, I suggest we revive this issue. As a starting point I have created a couple of plots in a [notebook](https://github.com/fliem/freesurfer/blob/qc_nb/notebooks/fs_qc_plots.ipynb).
What do you think?

**Comment by chrisgorgo** [comment]
This looks great - looping in @effigies who did some surface QC for fmriprep. This is how it looks:

![image](https://cloud.githubusercontent.com/assets/238759/23581651/aa358760-00cc-11e7-8b54-937c066ea20a.png)

**Comment by alexlicohen** [comment]
I also came across this: https://github.com/akeshavan/mindcontrol

Which might be work a look or forking some ideas from...

-Alex

**Comment by fliem** [comment]
Those look very pretty. @effigies could you please point me towards that fmriprep code. Did you do other things for qc as well (like calculating metrics...)? Thanks.

**Comment by fliem** [comment]
I have refined the [surface and subcortical plots](https://github.com/fliem/freesurfer/blob/qc_nb/notebooks/fs_qc_plots.ipynb), borrowing heavily from nilearn/niworkflows.
Next, I'll look into how to assemble them in htmls

**Comment by effigies** [comment]
Hi @fliem sorry that I totally missed this thread.

It looks like you found the visualization bit in niworkflows. We haven't really done any metrics at this point, though; we've [reconstructed surfaces using ANTs skull-stripping](https://github.com/poldracklab/fmriprep/blob/2b6f42e01313f25dc81b51d6331354203b59d798/fmriprep/workflows/anatomical.py#L82-L193) instead of `mri_watershed`, and are now [using `bbregister`](https://github.com/poldracklab/fmriprep/blob/2b6f42e01313f25dc81b51d6331354203b59d798/fmriprep/workflows/epi.py#L117-L145) to perform functional alignment. At present, my focus (poldracklab/fmriprep#391) is on sampling functional data to the surface and ensuring that these surfaces align properly in native T1 space, so that they can be used smoothly in non-FreeSurfer tools.

**Comment by fliem** [comment]
I have created first examples for single subjects and group html [reports](https://github.com/fliem/freesurfer/tree/qc_nb/example_reports/00_qc). Any suggestions?

**Comment by chrisgorgo** [comment]
They look great, although the subcortical segmentation seem a little
misaligned.

I wonder if this would make more sense as a standalone python module and
command line tool that could be reused in other projects.

**Comment by effigies** [comment]
I can't see the parcellation reports at all, either through Github or pulling locally and viewing...

But I do like @chrisfilo's idea of a separate module that would ease re-use in other projects. Since the base machinery for this is niworkflows-based, I think it'd be reasonable to pull this into niworkflows, itself. But I'd also entirely support an independent module that we then incorporate into `ReconAllRPT`.

**Comment by fliem** [comment]
* re misalignment: this was partially due to a bug in the code. However after fixing this, the contours still don't align perfectly. I have reverted to ROI overlays for now.
![subcort_contours1](https://cloud.githubusercontent.com/assets/5772811/25900037/57f2de4e-3592-11e7-832f-6d667897e7de.png)

* new example reports are [here](https://github.com/fliem/freesurfer/blob/qc_nb/example_reports/00_qc)

* Due to a [nilearn issue](https://github.com/nilearn/nilearn/issues/1450) with `.add_contours`, the white/pial-surface-plots show smaller brains than the subcortical overlays.

* Making it an own module or part of niworkflows sounds good to me. For the time being, I have created a command line tool as part of my freesurfer repo, and a docker container. If anyone wants to try it:
```
docker run --rm -ti \
-v /data/sourcedata:/data/in \
-v /data/freesurfer:/data/out \
--entrypoint=/code/run_qc.py \
fliem/freesurfer:qc_nb \
/data/in /data/out participant
```

**Comment by alexlicohen** [comment]
To view these without cloning, use rawgit.com

https://cdn.rawgit.com/fliem/freesurfer/qc_nb/example_reports/00_qc/freesurfer_qc_group_report.html

This looks pretty nice actually for scanning for large-scale FS failures.
(For more detailed screening I think we'll still need the more detailed
manual review and/or quantitative assessment like mindcontrol or mriqc

-Alex


----------------------------------------------------------------
Alexander Li Cohen, M.D., Ph.D.
E-mail: Alexander.Cohen2@childrens.harvard.edu (Medical/Science Email)
E-mail: alexcohen@gmail.com (Lifetime Email)
----------------------------------------------------------------

**Comment by satra** [comment]
@akeshavan - perhaps this can somehow be integrated with mindcontrol.

**Comment by effigies** [comment]
To be clear, the parcellation images are missing for everyone else, right? I'm seeing surfaces and segmentation only.

**Comment by alexlicohen** [comment]
Yes, I'm not seeing the cortical parcellations either.

-Alex


----------------------------------------------------------------
Alexander Li Cohen, M.D., Ph.D.
E-mail: Alexander.Cohen2@childrens.harvard.edu (Medical/Science Email)
E-mail: alexcohen@gmail.com (Lifetime Email)
----------------------------------------------------------------

On Wed, May 10, 2017 at 11:30 AM, Chris Markiewicz <notifications@github.com

**Comment by fliem** [comment]
Sorry, I can't figure out why the parcellations won't show. Here's one of them:
![sub-01_parc_lh_lateral](https://cloud.githubusercontent.com/assets/5772811/25913403/df48a162-35ba-11e7-8248-cd48c55f8fbe.png)

**Comment by PeerHerholz** [comment]
Ahoi hoi gang,

should we revive this given the recent developments over at `niworkflows` ?

**Comment by satra** [comment]
this is an entire segment of our current nobrainer collaboration with the freesurfer folks to develop autoqc methods in general and with the short term goal for freesurfer specifically. initial goal discard samples that will not work "well" with freesurfer, medium term goal - modify the samples to run through freesurfer - long term goal make freesurfer faster and more robust to inputs.

this doesn't mean anybody should stop trying to improve manual quality verification. this will also be used in our distributed training mode, where freesurfer runs locally and the lab can help provide feedback as to the quality .

also kwyk can be used in the short term to generate a quality assessment metric for T1 scans in 30s on a GPU.

btw, if anyone is interested in joining the autoqc effort just drop me a line and i will keep you informed about when we meet next (likely in jan, i'm mostly offline over the next few weeks).

**Comment by PeerHerholz** [comment]
Thx for the update and pointers @satra! 
I'm all for automatization and happy to not do QC manually (other than to train and inform a model), hehe. 

Re the things in this thread: as far as I understood it, the idea would be to generate a `BIDS App report` (comparable to `fmriprep` and `mriqc`) focusing on `FreeSurfer output` QC, right? If so, the information generated through the  `autoqc methods` could be included in this report besides some informative graphics, eh?

**Comment by satra** [comment]
> he idea would be to generate a BIDS App report (comparable to fmriprep and mriqc) focusing on FreeSurfer output QC, right? 

yes, and any useful thing created for the report itself could be used to improve autoqc.

> If so, the information generated through the autoqc methods could be included in this report besides some informative graphics, eh?

yes. the autoqc stuff can benefit from the report and in turn benefit the report in the future.

**Comment by Shotgunosine** [comment]
This is an old issue, but I think outputting QA reports would still be a big improvement.

**Comment by Remi-Gau** [comment]
note that for QA based on freesurfer output there is this: https://github.com/Qoala-T/QC
